lib/patch-by-id.js

```diff
@@ -0,0 +1,187 @@
+
+    if (!isOriginalQuery) {
+      var err = new Error('ctx.where:' + JSON.stringify(ctx.where) + ' is not valid; the `' +
+        idPropertyName + '` original value is ' + id);
+      err.statusCode = 403;
+      process.nextTick(function() {
+        cb(err);
+      });
+      return cb.promise;
+    }
+
+    if (strict) {
+      var instInfo = { '__unknownProperties': [] };
+      dao.applyStrictCheck(Model, strict, data,
+        instInfo, validateAndCallConnector);
+    } else {
+      validateAndCallConnector();
+    }
+
+    function validateAndCallConnector(err) {
+      if (err) return cb(err);
+      if (options.validate === false) {
+        return doUpdate(ctx.where, ctx.data);
+      }
+
+      // only when options.validate is not set, take model-setting into consideration
+      if (options.validate === undefined && Model.settings.automaticValidation === false) {
+        return doUpdate(ctx.where, ctx.data);
+      }
+
+      // Make sure id is not changed in the context of `before save`
+      checkId(ctx.data[idPropertyName], id, idPropertyName, cb);
+
+      dataObj = new Model(ctx.data);
+      dataObj.unsetAttribute(idPropertyName);
+      // validation required
+      dataObj.isValid(function(valid) {
+        if (valid) {
+          doUpdate(ctx.where, ctx.data);
+        } else {
+          cb(new ValidationError(dataObj), dataObj);
+        }
+      });
+    }
+  });
+
+  function doUpdate(where, data) {
+    try {
+      data = removeUndefined(data);
+      data = Model._coerce(data);
+    } catch (err) {
+      return process.nextTick(function() {
+        cb(err);
+      });
+    }
+
+    var ctx = {
+      Model: Model,
+      where: where,
+      data: data,
+      isNewInstance: false,
+      hookState: hookState,
+      options: options,
+    };
+    Model.notifyObserversOf('persist', ctx, function(err) {
+      if (err) return cb (err);
+      if (connector.update.length === 5) {
+        connector.updateAttributes(Model.modelName, where[idPropertyName], ctx.data, options, patchCallback);
+      } else {
+        connector.updateAttributes(Model.modelName, where[idPropertyName], ctx.data, patchCallback);
+      }
+    });
+
+    function patchCallback(err, info) {
+      if (err) return cb(err);
+      var ctx = {
+        Model: Model,
+        data: data,
+        hookState: hookState,
+        options: options,
+      };
+      Model.notifyObserversOf('loaded', ctx, function(err) {
+        if (err) return cb(err);
+        var context = {
+          Model: Model,
+          isNewInstance: false,
+          hookState: hookState,
+          options: options,
+        };
+        Model.notifyObserversOf('after save', context, function(err) {
+          if (!err) Model.emit('changed', info);
+          // Just for consistency we return { count: 1} for all connectors
+          cb(err, { count: 1 });
+        });
+      });
+    }
+  }
+  return cb.promise;
+};
+
+var idValidationFailure;
```

Comment:
I find this suspicious, isn't this a module-global singleton shared by all invocations of `checkId`? I don't think that will work.